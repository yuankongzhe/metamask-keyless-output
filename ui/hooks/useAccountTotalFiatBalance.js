import { shallowEqual, useSelector } from 'react-redux';
import {
  getAllTokens,
  getCurrentChainId,
  getCurrentCurrency,
  getMetaMaskCachedBalances,
  getTokenExchangeRates,
  getNativeCurrencyImage,
  getTokenList,
} from '../selectors';
import {
  getValueFromWeiHex,
  getWeiHexFromDecimalValue,
  sumDecimals,
} from '../../shared/modules/conversion.utils';
import {
  getConversionRate,
  getNativeCurrency,
} from '../ducks/metamask/metamask';
import { formatCurrency } from '../helpers/utils/confirm-tx.util';
import { getTokenFiatAmount } from '../helpers/utils/token-util';
import { isEqualCaseInsensitive } from '../../shared/modules/string-utils';
import { useTokenTracker } from './useTokenTracker';

export const useAccountTotalFiatBalance = (
  address,
  shouldHideZeroBalanceTokens,
) => {
  const currentChainId = useSelector(getCurrentChainId);
  const conversionRate = useSelector(getConversionRate);
  const currentCurrency = useSelector(getCurrentCurrency);

  const contractExchangeRates = useSelector(
    getTokenExchangeRates,
    shallowEqual,
  );

  const cachedBalances = useSelector(getMetaMaskCachedBalances);
  const balance = cachedBalances?.[address] ?? 0;
  const nativeFiat = getValueFromWeiHex({
    value: balance,
    toCurrency: currentCurrency,
    conversionRate,
    numberOfDecimals: 2,
  });

  const detectedTokens = useSelector(getAllTokens);
  const tokens = detectedTokens?.[currentChainId]?.[address] ?? [];
  // This selector returns all the tokens, we need it to get the image of token
  const allTokenList = useSelector(getTokenList);
  const allTokenListValues = Object.values(allTokenList);
  const primaryTokenImage = useSelector(getNativeCurrencyImage);
  const nativeCurrency = useSelector(getNativeCurrency);

  const { loading, tokensWithBalances } = useTokenTracker({
    tokens,
    address,
    includeFailedTokens: true,
    hideZeroBalanceTokens: shouldHideZeroBalanceTokens,
  });

  // Create fiat values for token balances
  const tokenFiatBalances = tokensWithBalances.map((token) => {
    const contractExchangeTokenKey = Object.keys(contractExchangeRates).find(
      (key) => isEqualCaseInsensitive(key, token.address),
    );
    const tokenExchangeRate =
      (contractExchangeTokenKey &&
        contractExchangeRates[contractExchangeTokenKey]) ??
      0;

    const totalFiatValue = getTokenFiatAmount(
      tokenExchangeRate,
      conversionRate,
      currentCurrency,
      token.string,
      token.symbol,
      false,
      false,
    );

    return totalFiatValue;
  });

  // Create an object with native token info. NOTE: Native token info is fetched from a separate controller
  const nativeTokenValues = {
    iconUrl: primaryTokenImage,
    symbol: nativeCurrency,
    fiatBalance: nativeFiat,
  };

  // To match the list of detected tokens with the entire token list to find the image for tokens
  const findMatchingTokens = (array1, array2) => {
    const result = [];

    array2.forEach((token2) => {
      const matchingToken = array1.find(
        (token1) => token1.symbol === token2.symbol,
      );

      if (matchingToken) {
        result.push({
          ...matchingToken,
          balance: token2.balance,
          string: token2.string,
          balanceError: token2.balanceError,
        });
      }
    });

    return result;
  };

  const matchingTokens = findMatchingTokens(
    allTokenListValues,
    tokensWithBalances,
  );

  // Combine native token, detected token with image in an array
  const allTokensWithFiatValues = [
    nativeTokenValues,
    ...matchingTokens.map((item, index) => ({
      ...item,
      fiatBalance: tokenFiatBalances[index],
    })),
  ];

  // Order of the tokens in this array is in decreasing order based on their fiatBalance
  const orderedTokenList = allTokensWithFiatValues.sort(
    (a, b) => parseFloat(b.fiatBalance) - parseFloat(a.fiatBalance),
  );

  // Total native and token fiat balance as a string (ex: "8.90")
  const totalFiatBalance = sumDecimals(
    nativeFiat,
    ...tokenFiatBalances,
  ).toString(10);

  // Fiat balance formatted in user's desired currency (ex: "$8.90")
  const formattedFiat = formatCurrency(totalFiatBalance, currentCurrency);

  // WEI Number which can be used with UserPreferencedCurrencyDisplay component
  let totalWeiBalance = getWeiHexFromDecimalValue({
    value: totalFiatBalance,
    fromCurrency: currentCurrency,
    conversionRate,
    invertConversionRate: true,
  });

  // If we have a totalFiatBalance of "0" and conversionRate of "0",
  // getWeiHexFromDecimalValue responds with "NaN"
  if (totalWeiBalance === 'NaN') {
    totalWeiBalance = '0x0';
  }

  return {
    formattedFiat,
    totalWeiBalance,
    totalFiatBalance,
    tokensWithBalances,
    loading,
    orderedTokenList,
  };
};
