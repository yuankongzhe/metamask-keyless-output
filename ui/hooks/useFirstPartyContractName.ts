import { NameType } from '@metamask/name-controller';
import { useSelector } from 'react-redux';
import { getCurrentChainId } from '../selectors';
import { FIRST_PARTY_CONTRACT_NAMES } from '../../shared/constants/first-party-contracts';

export function useFirstPartyContractName(
  value: string,
  type: NameType,
  variation?: string,
): string | null {
  if (type !== NameType.ETHEREUM_ADDRESS) {
    return null;
  }
  const currentChainId = useSelector(getCurrentChainId);
  const chainId = variation ?? currentChainId;
  const normalizedValue = value.toLowerCase();

  return (
    Object.keys(FIRST_PARTY_CONTRACT_NAMES).find(
      (name) =>
        FIRST_PARTY_CONTRACT_NAMES[name]?.[chainId]?.toLowerCase() ===
        normalizedValue,
    ) ?? null
  );
}
