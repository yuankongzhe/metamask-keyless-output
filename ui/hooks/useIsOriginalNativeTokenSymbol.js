import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import fetchWithCache from '../../shared/lib/fetch-with-cache';
import { CHAIN_ID_TO_CURRENCY_SYMBOL_MAP } from '../../shared/constants/network';
import { DAY } from '../../shared/constants/time';
import { useSafeChainsListValidationSelector } from '../selectors';

export function useIsOriginalNativeTokenSymbol(chainId, ticker, type) {
  const [isOriginalNativeSymbol, setIsOriginalNativeSymbol] = useState(null);
  const useSafeChainsListValidation = useSelector(
    useSafeChainsListValidationSelector,
  );
  useEffect(() => {
    async function getNativeTokenSymbol(networkId) {
      try {
        if (!useSafeChainsListValidation) {
          setIsOriginalNativeSymbol(true);
          return;
        }

        const mappedCurrencySymbol = CHAIN_ID_TO_CURRENCY_SYMBOL_MAP[chainId];
        if (mappedCurrencySymbol) {
          setIsOriginalNativeSymbol(mappedCurrencySymbol === ticker);
          return;
        }

        const safeChainsList = await fetchWithCache({
          url: 'https://chainid.network/chains.json',
          cacheOptions: { cacheRefreshTime: DAY },
          functionName: 'getSafeChainsList',
        });

        const matchedChain = safeChainsList.find(
          (network) => network.chainId === parseInt(networkId, 16),
        );

        const symbol = matchedChain?.nativeCurrency?.symbol ?? null;

        setIsOriginalNativeSymbol(symbol === ticker);
        return;
      } catch (err) {
        setIsOriginalNativeSymbol(false);
      }
    }

    getNativeTokenSymbol(chainId);
  }, [
    isOriginalNativeSymbol,
    chainId,
    ticker,
    type,
    useSafeChainsListValidation,
  ]);

  return isOriginalNativeSymbol;
}
