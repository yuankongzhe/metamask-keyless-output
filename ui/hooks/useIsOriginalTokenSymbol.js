import { useEffect, useState } from 'react';
import { getTokenSymbol } from '../store/actions';

export function useIsOriginalTokenSymbol(tokenAddress, tokenSymbol) {
  const [isOriginalNativeSymbol, setIsOriginalNativeSymbol] = useState(null);

  useEffect(() => {
    async function getTokenSymbolForToken(address) {
      const symbol = await getTokenSymbol(address);
      setIsOriginalNativeSymbol(symbol === tokenSymbol);
    }
    getTokenSymbolForToken(tokenAddress);
  }, [isOriginalNativeSymbol, tokenAddress, tokenSymbol]);

  return isOriginalNativeSymbol;
}
