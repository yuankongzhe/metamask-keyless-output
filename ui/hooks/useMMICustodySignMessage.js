import { useDispatch, useSelector } from 'react-redux';
import { mmiActionsFactory } from '../store/institutional/institution-background';
import { getAccountType } from '../selectors';
import {
  resolvePendingApproval,
  completedTx,
  showModal,
} from '../store/actions';

export function useMMICustodySignMessage() {
  const dispatch = useDispatch();
  const mmiActions = mmiActionsFactory();
  const accountType = useSelector(getAccountType);

  const custodySignFn = async (_msgData) => {
    if (accountType === 'custody') {
      try {
        await dispatch(resolvePendingApproval(_msgData.id));
        completedTx(_msgData.id);
        await dispatch(mmiActions.setWaitForConfirmDeepLinkDialog(true));
      } catch (err) {
        await dispatch(mmiActions.setWaitForConfirmDeepLinkDialog(true));
        await dispatch(
          showModal({
            name: 'TRANSACTION_FAILED',
            errorMessage: err.message,
            closeNotification: true,
            operationFailed: true,
          }),
        );
      }
    } else {
      // Non Custody accounts follow normal flow
      await dispatch(resolvePendingApproval(_msgData.id));
      completedTx(_msgData.id);
    }
  };

  return { custodySignFn };
}
