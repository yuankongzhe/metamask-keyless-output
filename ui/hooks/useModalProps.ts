import { useSelector, useDispatch } from 'react-redux';
import { hideModal } from '../store/actions';

interface ModalProps {
  props: Record<string, any>;
  hideModal: () => void;
}

export function useModalProps(): ModalProps {
  const modalProps = useSelector((state: any) => {
    return state.appState.modal.modalState?.props;
  });

  const dispatch = useDispatch();
  const onHideModal = () => dispatch(hideModal());

  return { props: modalProps, hideModal: onHideModal };
}
