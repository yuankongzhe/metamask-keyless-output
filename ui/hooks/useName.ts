import {
  FALLBACK_VARIATION,
  NameEntry,
  NameType,
} from '@metamask/name-controller';
import { useSelector } from 'react-redux';
import { isEqual } from 'lodash';
import { getCurrentChainId, getNames } from '../selectors';

function normalizeValue(value: string, type: string): string {
  switch (type) {
    case NameType.ETHEREUM_ADDRESS:
      return value.toLowerCase();

    default:
      return value;
  }
}

function getVariationKey(type: string, chainId: string): string {
  switch (type) {
    case NameType.ETHEREUM_ADDRESS:
      return chainId;

    default:
      return '';
  }
}

export function useName(
  value: string,
  type: NameType,
  variation?: string,
): NameEntry {
  const names = useSelector(getNames, isEqual);
  const chainId = useSelector(getCurrentChainId);
  const normalizedValue = normalizeValue(value, type);
  const typeVariationKey = getVariationKey(type, chainId);
  const variationKey = variation ?? typeVariationKey;
  const variationsToNameEntries = names[type]?.[normalizedValue] ?? {};

  const variationEntry = variationsToNameEntries[variationKey];
  const fallbackEntry = variationsToNameEntries[FALLBACK_VARIATION];

  const entry =
    !variationEntry?.name && fallbackEntry
      ? fallbackEntry
      : variationEntry ?? {};

  const {
    name = null,
    sourceId = null,
    origin = null,
    proposedNames = {},
  } = entry;
  return {
    name,
    sourceId,
    proposedNames,
    origin,
  };
}
