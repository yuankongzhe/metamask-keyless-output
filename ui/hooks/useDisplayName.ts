import { NameType } from '@metamask/name-controller';
import { useSelector } from 'react-redux';
import { getMemoizedMetadataContractName } from '../selectors';
import { useName } from './useName';
import { useFirstPartyContractName } from './useFirstPartyContractName';

/**
 * Attempts to resolve the name for the given parameters.
 *
 * @param value
 * @param type
 * @param variation
 * @returns An object with two properties:
 * - `name` {string|null} - The display name, if it can be resolved, otherwise null.
 * - `hasPetname` {boolean} - True if there is a petname for the given address.
 */
export function useDisplayName(
  value: string,
  type: NameType,
  variation?: string,
): { name: string | null; hasPetname: boolean } {
  const nameEntry = useName(value, type, variation);
  const firstPartyContractName = useFirstPartyContractName(
    value,
    type,
    variation,
  );
  const contractName = useSelector((state) =>
    (getMemoizedMetadataContractName as any)(state, value),
  );
  return {
    name: nameEntry?.name || firstPartyContractName || contractName || null,
    hasPetname: Boolean(nameEntry?.name),
  };
}
