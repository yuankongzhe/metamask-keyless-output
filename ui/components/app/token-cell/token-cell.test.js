import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { fireEvent } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import { useTokenFiatAmount } from '../../../hooks/useTokenFiatAmount';

import { useIsOriginalTokenSymbol } from '../../../hooks/useIsOriginalTokenSymbol';
import TokenCell from '.';

jest.mock('react-redux', () => {
  const actual = jest.requireActual('react-redux');

  return {
    ...actual,
    useSelector: jest.fn(),
  };
});

jest.mock('../../../hooks/useTokenFiatAmount', () => {
  return {
    useTokenFiatAmount: jest.fn(),
  };
});

jest.mock('../../../hooks/useIsOriginalTokenSymbol', () => {
  return {
    useIsOriginalTokenSymbol: jest.fn(),
  };
});
describe('Token Cell', () => {
  const mockState = {
    metamask: {
      selectedAddress: '0xAddress',
      contractExchangeRates: {
        '0xAnotherToken': 0.015,
      },
      currentCurrency: 'usd',
      currencyRates: {
        ETH: {
          conversionRate: 7.0,
        },
      },
      preferences: {},
      providerConfig: {
        chainId: '0x1',
        ticker: 'ETH',
        type: 'mainnet',
      },
    },
  };

  useIsOriginalTokenSymbol.mockReturnValue(true);

  // two tokens with the same symbol but different addresses
  const MOCK_GET_TOKEN_LIST = {
    '0xAddress': {
      name: 'TEST-2',
      erc20: true,
      symbol: 'TEST',
      decimals: 18,
      address: '0xAddress',
      iconUrl: './images/test_1_image.svg',
      aggregators: [],
    },
    '0xAnotherToken': {
      name: 'TEST',
      erc20: true,
      symbol: 'TEST',
      decimals: 18,
      address: '0xANoTherToKen',
      iconUrl: './images/test_image.svg',
      aggregators: [],
    },
  };

  const mockStore = configureMockStore([thunk])(mockState);

  const props = {
    address: '0xAnotherToken',
    symbol: 'TEST',
    string: '5.000',
    currentCurrency: 'usd',
    onClick: jest.fn(),
  };

  useSelector.mockReturnValue(MOCK_GET_TOKEN_LIST);
  useTokenFiatAmount.mockReturnValue('5.00');

  it('should match snapshot', () => {
    const { container } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  it('calls onClick when clicked', () => {
    const { queryByTestId } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    fireEvent.click(queryByTestId('multichain-token-list-button'));

    expect(props.onClick).toHaveBeenCalled();
  });

  it('should render the correct token and filter by symbol and address', () => {
    const { getByTestId, getByAltText } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    const image = getByAltText('TEST logo');

    expect(getByTestId('multichain-token-list-item-value')).toBeInTheDocument();
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', './images/test_image.svg');
  });
});
