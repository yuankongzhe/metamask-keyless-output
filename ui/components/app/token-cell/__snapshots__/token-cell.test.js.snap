// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token Cell should match snapshot 1`] = `
<div>
  <div
    class="mm-box multichain-token-list-item mm-box--display-flex mm-box--gap-4 mm-box--flex-direction-column"
    data-testid="multichain-token-list-item"
  >
    <a
      class="mm-box multichain-token-list-item__container-cell mm-box--padding-4 mm-box--display-flex mm-box--flex-direction-row"
      data-testid="multichain-token-list-button"
      href="#"
    >
      <div
        class="mm-box mm-badge-wrapper mm-box--margin-right-3 mm-box--display-inline-block"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-token mm-avatar-token--with-halo mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
        >
          <img
            aria-hidden="true"
            class="mm-avatar-token__token-image--blurred"
            src="./images/test_image.svg"
          />
          <img
            alt="TEST logo"
            class="mm-avatar-token__token-image--size-reduced"
            src="./images/test_image.svg"
          />
        </div>
        <div
          class="mm-box mm-badge-wrapper__badge-container mm-badge-wrapper__badge-container--circular-top-right"
        >
          <div
            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--rounded-full mm-box--border-color-border-muted box--border-style-solid box--border-width-1"
          >
            ?
          </div>
        </div>
      </div>
      <div
        class="mm-box multichain-token-list-item__container-cell--text-container mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
        style="flex-grow: 1; overflow: hidden;"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        >
          <div
            class="mm-box mm-box--display-inline-block mm-box--width-1/3"
          >
            <span
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--ellipsis mm-box--color-text-default"
            >
              TEST
            </span>
          </div>
          <p
            class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--text-align-end mm-box--width-2/3 mm-box--color-text-default"
            data-testid="multichain-token-list-item-secondary-value"
          >
            5.00
          </p>
        </div>
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        >
          <div
            class="mm-box mm-box--width-1/3"
          >
            <p
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--ellipsis mm-box--color-text-alternative"
              data-testid="multichain-token-list-item-token-name"
            >
              TEST
            </p>
          </div>
          <div
            class="mm-box mm-box--width-2/3"
            style="overflow: hidden;"
          >
            <p
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--text-align-end mm-box--color-text-alternative"
              data-testid="multichain-token-list-item-value"
            >
              5.000
               
              TEST
            </p>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
